pkg/resource/api.go

```diff
@@ -63,41 +67,119 @@ func (a *APIManagedConnectionPropagator) PropagateConnection(ctx context.Context
 		Namespace: mg.GetWriteConnectionSecretToReference().Namespace,
 		Name:      mg.GetWriteConnectionSecretToReference().Name,
 	}
-	mgcs := &corev1.Secret{}
-	if err := a.client.Get(ctx, n, mgcs); err != nil {
+	from := &corev1.Secret{}
+	if err := a.client.Get(ctx, n, from); err != nil {
 		return errors.Wrap(err, errGetSecret)
 	}
 
 	// Make sure the managed resource is the controller of the connection secret
 	// it references before we propagate it. This ensures a managed resource
 	// cannot use Crossplane to circumvent RBAC by propagating a secret it does
 	// not own.
-	if c := metav1.GetControllerOf(mgcs); c == nil || c.UID != mg.GetUID() {
+	if c := metav1.GetControllerOf(from); c == nil || c.UID != mg.GetUID() {
 		return errors.New(errSecretConflict)
 	}
 
-	cmcs := LocalConnectionSecretFor(o, MustGetKind(o, a.typer))
-	if _, err := util.CreateOrUpdate(ctx, a.client, cmcs, func() error {
-		// Inside this anonymous function cmcs could either be unchanged (if
-		// it does not exist in the API server) or updated to reflect its
-		// current state according to the API server.
-		if c := metav1.GetControllerOf(cmcs); c == nil || c.UID != o.GetUID() {
-			return errors.New(errSecretConflict)
-		}
-		cmcs.Data = mgcs.Data
-		meta.AddAnnotations(cmcs, map[string]string{
-			AnnotationKeyPropagateFromNamespace: mgcs.GetNamespace(),
-			AnnotationKeyPropagateFromName:      mgcs.GetName(),
-			AnnotationKeyPropagateFromUID:       string(mgcs.GetUID()),
-		})
-		return nil
-	}); err != nil {
+	to := LocalConnectionSecretFor(o, MustGetKind(o, a.typer))
```

Comment:
I think the returned secret from `LocalConnectionSecretFor` and `ConnectionSecretFor` should be of type `connection.crossplane.io/v1alpha1` for all this to work.

Reply:
Ah, good catch. I did that but it seems to have been lost somewhere along the way.